tensorflow/core/kernels/one_hot_op.cc

```diff
@@ -113,44 +113,68 @@ class OneHotOp : public OpKernel {
   TF_DISALLOW_COPY_AND_ASSIGN(OneHotOp);
 };
 
-#define REGISTER_ONE_HOT(type)                           \
-  REGISTER_KERNEL_BUILDER(Name("OneHot")                 \
-                              .Device(DEVICE_CPU)        \
-                              .TypeConstraint<type>("T") \
-                              .HostMemory("depth"),      \
-                          OneHotOp<CPUDevice, type>);
+#define REGISTER_ONE_HOT_INDEX(type, index_type)                  \
+  REGISTER_KERNEL_BUILDER(Name("OneHot")                          \
+                              .Device(DEVICE_CPU)                 \
+                              .TypeConstraint<index_type>("TI")   \
+                              .TypeConstraint<type>("T")          \
+                              .HostMemory("depth"),               \
+                          OneHotOp<CPUDevice, type, index_type>);
 
-TF_CALL_NUMBER_TYPES(REGISTER_ONE_HOT);
+#define REGISTER_ONE_HOT(type)          \
+  REGISTER_ONE_HOT_INDEX(type, int8);   \
+  REGISTER_ONE_HOT_INDEX(type, int16);  \
+  REGISTER_ONE_HOT_INDEX(type, int32);  \
+  REGISTER_ONE_HOT_INDEX(type, int64);  \
+  REGISTER_ONE_HOT_INDEX(type, uint8)
+
+TF_CALL_ALL_TYPES(REGISTER_ONE_HOT);
 
 #if GOOGLE_CUDA
 
 // Forward declarations of the functor specializations for GPU.
 namespace functor {
-#define DECLARE_GPU_SPEC(T)                                                   \
-  template <>                                                                 \
-  void OneHot<GPUDevice, T>::Compute(                                         \
-      const GPUDevice& d, const typename TTypes<int64>::ConstMatrix& indices, \
-      const typename TTypes<T>::ConstScalar& on_value,                        \
-      const typename TTypes<T>::ConstScalar& off_value,                       \
-      typename TTypes<T, 3>::Tensor* output);                                 \
-  extern template struct OneHot<GPUDevice, T>;
+#define DECLARE_GPU_SPEC_INDEX(T, TI)                                       \
+  template <>                                                               \
+  void OneHot<GPUDevice, T, TI>::Compute(                                   \
+      const GPUDevice& d, const typename TTypes<TI>::ConstMatrix& indices,  \
+      const typename TTypes<T>::ConstScalar& on_value,                      \
+      const typename TTypes<T>::ConstScalar& off_value,                     \
+      typename TTypes<T, 3>::Tensor* output);                               \
+  extern template struct OneHot<GPUDevice, T, TI>;
+
+#define DECLARE_GPU_SPEC(T)          \
+  DECLARE_GPU_SPEC_INDEX(T, int8);   \
```

Comment:
we only need int32 and int64.  no one uses indexing with other types and this bloats the C++ objects (every instantiation of a template creates a complete copy of the code, * all the types)